### PR TITLE
Initial neo-vm support

### PR DESCRIPTION
**Pull request by osmirnov**
I've started the implementation of NeoVM support. My goal not only make it running but generalize the API and adapt HyperVM later.

### CODE REVIEW

File: src/NeoSharp.VM.NeoVM/StackItemType.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿namespace NeoSharp.VM.NeoVM
+{
+    internal enum StackItemType : byte
```
Comment (shargon) [inline src/NeoSharp.VM.NeoVM/StackItemType.cs]: This enum is already on the nuget Package ( NeoSharp.VM.Interop )

File: src/NeoSharp.VM.NeoVM/StackAccessor.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+using Neo.VM;
+using NeoSharp.Core.Extensions;
+using NeoSharp.Core.VM;
+using VMArray = Neo.VM.Types.Array;
+
+namespace NeoSharp.VM.NeoVM
+{
+    public class StackAccessor : IStackAccessor
+    {
+        private readonly RandomAccessStack<StackItem> _stack;
```
Comment (shargon) [inline src/NeoSharp.VM.NeoVM/StackAccessor.cs]: We need to dispose the items when is not used, native VM require this

File: src/NeoSharp.Core/SmartContract/TriggerType.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿namespace NeoSharp.Core.SmartContract
+{
+    public enum TriggerType : byte
+    {
```
Comment (shargon) [inline src/NeoSharp.Core/SmartContract/TriggerType.cs]: This enum is already on the nuget Package ( NeoSharp.VM.Interop )

File: src/NeoSharp.Core/Blockchain/Processing/InvocationTransactionPersister2.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
+using System.Threading.Tasks;
+using NeoSharp.Core.Models;
+using NeoSharp.Core.SmartContract;
+
+namespace NeoSharp.Core.Blockchain.Processing
+{
+    public class InvocationTransactionPersister2: ITransactionPersister<InvocationTransaction>
+    {
```
Comment (shargon) [inline src/NeoSharp.Core/Blockchain/Processing/InvocationTransactionPersister2.cs]: Is this class used?

File: src/NeoSharp.Core/Blockchain/Processing/InvocationTransactionPersister.cs [hunk 5]
```diff
@@ -1,14 +1,74 @@
...
     public class InvocationTransactionPersister: ITransactionPersister<InvocationTransaction>
     {
-        public Task Persist(InvocationTransaction invocationTx)
+        private readonly IVMFactory _vmFactory;
+        private readonly IBinarySerializer _binarySerializer;
+        private readonly IContainer _container;
+
+        public InvocationTransactionPersister(
+            IVMFactory vmFactory,
+            IBinarySerializer binarySerializer,
```
Comment (shargon) [inline src/NeoSharp.Core/Blockchain/Processing/InvocationTransactionPersister.cs]: We use the serializer is used in static way

File: src/NeoSharp.Core/VM/IStackAccessor.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
+﻿using System.Collections;
+using System.Numerics;
+
+namespace NeoSharp.Core.VM
+{
+    public interface IStackAccessor
+    {
```
Comment (shargon) [inline src/NeoSharp.Core/VM/IStackAccessor.cs:7]: Is this interface needed? Then should be moved to the `NeoSharp.VM.Interop` package

File: src/NeoSharp.Core/VM/StateMachine.cs [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+            }
+
+            stack.Push(asset.Expiration);
+
+            return true;
+        }
+
+        private bool Contract_Create(IStackAccessor stack)
+        {
+            var script = stack.PopByteArray();
```
Comment (shargon) [inline src/NeoSharp.Core/VM/StateMachine.cs:160]: This method call `Dispose` to the base StackItem?

File: src/NeoSharp.VM.NeoVM/CryptoAdapter.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
+﻿using Neo.VM;
+using NeoSharp.Cryptography;
+
+namespace NeoSharp.VM.NeoVM
+{
+    public class CryptoAdapter : ICrypto
+    {
```
Comment (shargon) [inline src/NeoSharp.VM.NeoVM/CryptoAdapter.cs:7]: We need this class?

File: src/NeoSharp.VM.NeoVM/IApplicationEngine.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
+﻿using Neo.VM;
+using NeoSharp.Types;
+
+namespace NeoSharp.VM.NeoVM
+{
+    public interface IApplicationEngine
+    {
```
Comment (shargon) [inline src/NeoSharp.VM.NeoVM/IApplicationEngine.cs]: Should implement the same interface between the VM, defined already in NeoSharp.VM.Interop

**Review by shargon — COMMENTED** [review]
I think that both VM should share the same interfaces, defined in NeoSharp.VM.Interop.

**Comment by shargon** [comment]
@osmirnov there are problems with items like interopStackItem, i will deal with it on :

- https://github.com/neo-project/neo-vm/pull/64
- https://github.com/neo-project/neo-vm/pull/66

**Comment by shargon** [comment]
https://github.com/neo-project/neo-vm/pull/67 this solution will help with this issues

**Comment by shargon** [comment]
@osmirnov please review the last changes, is using a compiled version of NeoVM with the last changes.

File: src/NeoSharp.VM/Types/InteropService.cs [hunk 10]
```diff
@@ -88,84 +77,70 @@ public InteropService()
...
+        /// <returns>Return false if something is wrong</returns>
         public bool Invoke(byte[] method, IExecutionEngine engine)
         {
-            uint hash = method.Length == 4 ? BitConverter.ToUInt32(method, 0)
-                 : Encoding.ASCII.GetString(method).ToInteropMethodHash();
+            var hash = method.Length == 4
+                ? BitConverter.ToUInt32(method, 0)
+                : GetMethodHash(Encoding.ASCII.GetString(method));
+
+            void InvokeOnSysCall(SysCallResult result) => OnSysCall?.Invoke(this, new SysCallArgs(engine, GetMethodName(hash), hash, result));
```
Comment (shargon) [inline src/NeoSharp.VM/Types/InteropService.cs:118]: This is very weird xD

**Comment by shargon** [comment]
Execution successful of virtual smart contract

```
virtual contract add 0x0066
testinvoke 0x9e5bf7f291e4d8edd25e7b3c790c6f11590f9754 Application
```

image

**Comment by shargon** [comment]
Now is used the nuGet package, Erik released this version :)

**Review by shargon — APPROVED** [review]

**Comment by shargon** [comment]
@osmirnov do you know why travis is throwing an exception?

File: src/NeoSharp.VM.NeoVM/ApplicationEngine.cs [hunk 11]
```diff
@@ -0,0 +1,676 @@
...
+        public ApplicationEngine(ExecutionEngineArgs args) : base(args)
+        {
+            _isDisposed = false;
+            _gasConsumed = _gasAmount = 0;
+            _state = EVMState.None;
+            _scriptTable = new List<byte[]>();
+
+            _contractScriptTable = new ContractScriptTable(args.ScriptTable);
+
+            _engine = new ExecutionEngine(
```
Comment (osmirnov) [inline src/NeoSharp.VM.NeoVM/ApplicationEngine.cs:171]: @shargon it has to be disposed, right?

File: src/NeoSharp.Core/VM/IStackAccessor.cs (hunk 6 above)
Comment (osmirnov) [reply src/NeoSharp.Core/VM/IStackAccessor.cs:7]: I've realized it was probably not the best thing to introduce.

File: src/NeoSharp.Core/VM/StateMachine.cs (hunk 7 above)
Comment (osmirnov) [reply src/NeoSharp.Core/VM/StateMachine.cs:160]: I'll probably switch to VM abstractions instead. But yes it's a good point.

File: src/NeoSharp.VM/Types/InteropService.cs (hunk 10 above)
Comment (osmirnov) [reply src/NeoSharp.VM/Types/InteropService.cs:118]: Yep, I used to use `Func` or `Action` delegates instead but VS converts them into local methods all the time.

**Comment by osmirnov** [comment]
> @osmirnov do you know why travis is throwing an exception?

Yes, I think it is related to https://github.com/CityOfZion/neo-sharp/pull/548/files#diff-432d787ad9965e43ec651da1be56928dR80 not to this PR

File: src/NeoSharp.VM.NeoVM/ApplicationEngine.cs (hunk 11 above)
Comment (shargon) [reply src/NeoSharp.VM.NeoVM/ApplicationEngine.cs:171]: Is from NeoVM, so it is not necessary

**Pull request merged**